plutus-core/cost-model/budgeting-bench/Bench.hs

```diff
@@ -1,154 +1,254 @@
 
-type UntypedPlain f (uni :: GHC.Type -> GHC.Type) (fun :: GHC.Type) = f Name uni fun ()
+type PlainTerm = UPLC.Term Name DefaultUni DefaultFun ()
 
-runTermBench :: String -> UntypedPlain UT.Term DefaultUni DefaultFun -> Benchmark
+-- TODO.  I'm not totally sure what's going on here.  `env` is supposed to
+-- produce data that will be supplied to the things being benchmarked.  Here
+-- we've got a term and we evaluate it to get back the budget consumed, but then
+-- we throw that away and evaluate the term again.  This may have the effect of
+-- avoiding warmup, which could be a good thing.  Let's look into that.
+runTermBench :: String -> PlainTerm -> Benchmark
 runTermBench name term = env
     (do
         (_result, budget) <-
-          pure $ runCekNoEmit defBuiltinsRuntime Cek.counting term
+          pure $ (unsafeEvaluateCek defBuiltinsRuntime) term
         pure budget
         )
-    (\_ -> bench name $ nf (unsafeEvaluateCek defBuiltinsRuntime) term)
+    $ \_ -> bench name $ nf (unsafeEvaluateCek defBuiltinsRuntime) term
 
--- Copying the bytestring here, because otherwise it'll be exactly the same, and the equality will short-circuit.
-benchSameTwoByteStrings :: DefaultFun -> Benchmark
-benchSameTwoByteStrings name = createTwoTermBuiltinBench name (byteStringsToBench seedA)
-                               ((\(bs, e) -> (BS.copy bs, e)) <$> byteStringsToBench seedA)
+-- Create a term applying a builtin to one argument
+mkApp1 :: (DefaultUni `Includes` a) => DefaultFun -> a -> PlainTerm
+mkApp1 name x =
+    erase $ mkIterApp () (builtin () name) [mkConstant () x]
 
-benchTwoByteStrings :: DefaultFun -> Benchmark
-benchTwoByteStrings name = createTwoTermBuiltinBench name (byteStringsToBench seedA) (byteStringsToBench seedB)
+-- Create a term applying a builtin to two arguments
+mkApp2
+    :: (DefaultUni `Includes` a, DefaultUni `Includes` b)
+    =>  DefaultFun -> a -> b -> PlainTerm
+mkApp2 name x y =
+    erase $ mkIterApp () (builtin () name) [mkConstant () x,  mkConstant () y]
 
-benchBytestringOperations :: DefaultFun -> Benchmark -- TODO the numbers are a bit too big here
-benchBytestringOperations name = createTwoTermBuiltinBench @Integer @BS.ByteString name numbers (byteStringsToBench seedA)
-    where
-        numbers = expToBenchingInteger <$> expsToBench
+-- Create a term applying a builtin to three arguments
+mkApp3
+    :: (DefaultUni `Includes` a, DefaultUni `Includes` b, DefaultUni `Includes` c)
+    =>  DefaultFun -> a -> b -> c -> PlainTerm
+mkApp3 name x y z =
+    erase $ mkIterApp () (builtin () name) [mkConstant () x,  mkConstant () y, mkConstant () z]
 
+{- | Given a builtin function f of type a * b -> _ together with lists xs::[a] and
+   ys::[b] (along with their memory sizes), create a collection of benchmarks
+   which run f on all pairs in {(x,y}: x in xs, y in ys}. -}
+{- | The use of bgroups will cause Criterion to give the benchmarks names like
```

Comment:
was this meant to be one comment?

Reply:
Not really.  The second one applies to a number of functions, so I've moved it up to appear before them all.